### PR TITLE
QL: Reduce nesting of same bool queries

### LINKED ISSUE #96236: QL: Reduce number of nested bool queries
### Description

Certain OR expressions are translated into a number of nested bool queries which, unless balanced, end up being too deep and rejected by ES/Lucene.

Reproduced on 8.7.1 thanks to @e-parth-pathak:

1. Create an Index and ingest data:
```
PUT /library/_bulk?refresh
{"index":{"_id":"Leviathan Wakes"}}
{"name":"Leviathan Wakes","author":"James S.A. Corey","release_date":"2011-06-02","page_count":561}
{"index":{"_id":"Hyperion"}}
{"name":"Hyperion","author":"Dan Simmons","release_date":"1989-05-26","page_count":482}
{"index":{"_id":"Dune"}}
{"name":"Dune","author":"Frank Herbert","release_date":"1965-06-01","page_count":604}
{"index":{"_id":"Persuation"}}
{"name":"Persuation","author":"Robert Cialdini","release_date":"2023-06-01","page_count":304}
{"index":{"_id":"Sapiens"}}
{"name":"Sapiens","author":"Yuval Noah Harari","release_date":"1969-06-01","page_count":204}
{"index":{"_id":"Hooked"}}
{"name":"Hooked","author":"Nir Eyal","release_date":"2022-06-01","page_count":408}
```

2. Run a query:

```
POST /_sql/translate
{
  "query": "SELECT * FROM library WHERE name like '%nir%' OR name like '%yuval%' OR name like '%dan%' OR name like '%frank%' OR name like '%robert%' OR name like '%leviathan%'"
}
```

3. The response that I got:

```json
{
  "size": 1000,
  "query": {
    "bool": {
      "should": [
        {
          "bool": {
            "should": [
              {
                "bool": {
                  "should": [
                    {
                      "bool": {
                        "should": [
                          {
                            "bool": {
                              "should": [
                                {
                                  "wildcard": {
                                    "name.keyword": {
                                      "wildcard": "*nir*",
                                      "boost": 1
                                    }
                                  }
                                },
                                {
                                  "wildcard": {
                                    "name.keyword": {
                                      "wildcard": "*yuval*",
                                      "boost": 1
                                    }
                                  }
                                }
                              ],
                              "boost": 1
                            }
                          },
                          {
                            "wildcard": {
                              "name.keyword": {
                                "wildcard": "*dan*",
                                "boost": 1
                              }
                            }
                          }
                        ],
                        "boost": 1
                      }
                    },
                    {
                      "wildcard": {
                        "name.keyword": {
                          "wildcard": "*frank*",
                          "boost": 1
                        }
                      }
                    }
                  ],
                  "boost": 1
                }
              },
              {
                "wildcard": {
                  "name.keyword": {
                    "wildcard": "*robert*",
                    "boost": 1
                  }
                }
              }
            ],
            "boost": 1
          }
        },
        {
          "wildcard": {
            "name.keyword": {
              "wildcard": "*leviathan*",
              "boost": 1
            }
          }
        }
      ],
      "boost": 1
    }
  },
  "_source": false,
  "fields": [
    {
      "field": "author"
    },
    {
      "field": "name"
    },
    {
      "field": "page_count"
    },
    {
      "field": "release_date",
      "format": "strict_date_optional_time_nanos"
    }
  ],
  "sort": [
    {
      "_doc": {
        "order": "asc"
      }
    }
  ],
  "track_total_hits": -1
}
```

Note that while the lowest nested bool query uses two entries, the rest do not and instead rely on another bool wrapper.
Either all entries should be put into a should clause OR, all the OR entries need to be balanced and be wrapped on the same level.

P.S. This has been addressed in the past by #30267 however it still seems to occur when dealing with like comparisons.

### PR DESCRIPTION
**Pull request by costin**
Improve transpilation of chained OR and AND queries by replacing
 boolQuery wrapping with extension of the queries inside the bool clause

Fix #96236

### CODE REVIEW

**Review by luigidellaquila — APPROVED** [review]
LGTM
I'd just add one or two unit tests with multiple conditions and mixed AND/OR

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/querydsl/query/BoolQuery.java [hunk 1]
```diff
@@ -23,52 +29,59 @@ public class BoolQuery extends Query {
      * {@code true} for boolean {@code AND}, {@code false} for boolean {@code OR}.
      */
     private final boolean isAnd;
-    private final Query left;
-    private final Query right;
+    private final List<Query> queries;
```
Comment (astefan) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/querydsl/query/BoolQuery.java:32]: Imo, this change is a bit forced, conceptually speaking. A boolean query by definition is a boolean operation between two expressions, just like a math operation is. Having multiple expressions ALL linked together by either OR or AND seems like a specialized boolean query, MultiBoolQuery if you want. The BoolQueryBuilder in ES is built similarly, but is generalized: a list of must, a list of should, a list of filters etc.

I would have expected a solution around "flattening" the bool statements, but not touching the BoolQuery in this way.

File: x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/querydsl/query/BoolQueryTests.java [hunk 2]
```diff
@@ -35,15 +37,15 @@ public void testEqualsAndHashCode() {
     }
 
     private static BoolQuery copy(BoolQuery query) {
-        return new BoolQuery(query.source(), query.isAnd(), query.left(), query.right());
+        return new BoolQuery(query.source(), query.isAnd(), left(query), right(query));
     }
 
     private static BoolQuery mutate(BoolQuery query) {
         List<Function<BoolQuery, BoolQuery>> options = Arrays.asList(
```
Comment (astefan) [inline x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/querydsl/query/BoolQueryTests.java:44]: This test doesn't fully reflect the changes in this PR, but I'm ok with it.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/querydsl/query/BoolQuery.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/querydsl/query/BoolQuery.java:32]: Confirmed with the search team that the two types of queries are equivalent as Lucene normalizes them to the same query (the one created above) when `minimum_should_query` is the default (1 - which it is).

File: x-pack/plugin/eql/src/test/resources/querytranslator_tests.txt [hunk 3]
```diff
@@ -208,14 +208,14 @@ twoFunctionsEqualsBooleanLiterals-caseSensitive
 process where endsWith(process_path, "x") == true and endsWith(process_path, "yx") != true
 ;
 {"bool":{"must":[{"wildcard":{"process_path":{"wildcard":"*x","boost":1.0}}},
-{"bool":{"must_not":[{"wildcard":{"process_path":{"wildcard":"*yx","boost":1.0}}}],"boost":1.0}}],"boost":1.0}}
+{"bool":{"must_not":[{"wildcard":{"process_path":{"wildcard":"*yx","boost":1.0}}}],"boost":1.0}}
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/eql/src/test/resources/querytranslator_tests.txt:211]: Confirmed with the search team that the two types of queries are equivalent as Lucene normalizes them to the same query (the one created above) when `minimum_should_query` is the default (1 - which it is).

**Comment by costin** [comment]
> LGTM I'd just add one or two unit tests with multiple conditions and mixed AND/OR

I've added two more tests - OR OR AND OR OR plus (OR OR) AND (OR OR OR)

File: x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/querydsl/query/BoolQueryTests.java (hunk 2 above)
Comment (costin) [reply x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/querydsl/query/BoolQueryTests.java:44]: Fixed it nevertheless.

**Pull request merged**